lib/ts/recipe/session/recipe.ts

```diff
@@ -158,21 +153,53 @@ export default class Session extends RecipeModule<unknown, unknown, unknown, any
             });
         }
 
+        const globalValidators: SessionClaimValidator[] = await getGlobalClaimValidators();
+
         // We validate all the global claims
         const invalidClaims = await this.validateClaims({ userContext });
 
-        // Check if any of those claim errors requests a redirection
-        const invalidClaimRedirectPath = popInvalidClaimRedirectPathFromContext(userContext);
-        if (invalidClaims.length > 0 && invalidClaimRedirectPath !== undefined) {
+        if (invalidClaims.length > 0) {
             if (redirectInfo !== undefined) {
                 // if we have to redirect and we have success context we wanted to use we save it in localstorage
-                // this way after the other page did solved the validation error it can contine
+                // this way after the other page did solved the validation error it can continue
                 // the sign in process by calling this function without passing the redirect info
                 const jsonContext = JSON.stringify(redirectInfo);
                 await setLocalStorage("supertokens-success-redirection-context", jsonContext);
             }
-            // then we do the redirection.
-            return SuperTokens.getInstanceOrThrow().redirectToUrl(invalidClaimRedirectPath, history);
+
+            // we try to find claim validator among failed validators with onFailure cb that returns string
+            const getRedirectionPath = async () => {
+                const callbacks = [];
+                for (const claim of invalidClaims) {
+                    const validator = globalValidators.find(({ id }) => id === claim.validatorId);
+                    if (validator?.onFailureRedirection) {
+                        callbacks.push(validator?.onFailureRedirection?.());
+                    }
+                }
+                return (await Promise.all(callbacks)).filter(Boolean)[0];
+            };
+
+            const redirectPath = await getRedirectionPath();
```

Comment:
1. I don't see a reason why this is extracted into a function
2. We shouldn't call all of them at the same time: 
  a/ it might be misleading to users (if they return a value from it they could expect the redirection to happen)
  b/ it could be called in non-deterministic order, which may cause issues when testing in the future
  c/ there is no reason to do it: we only want the first one, and they are almost always sync functions anyway.

I think both this and the success part could be simplified into just a for loop (both could be done with reduce, but it's kind of awkward with promises)